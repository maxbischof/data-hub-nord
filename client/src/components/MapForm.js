import React, { useState } from 'react'
import styled from 'styled-components'
import Button from './ui/Button'
import { fetchAdress } from './../lib/geo'
import Bottleneck from 'bottleneck/es5'
import { useProgressStatus } from '../hooks/useProgressStatus'
import LoadingDots from './ui/LoadingDots'

export default function MapForm({ tableData, setMapData, columnNames }) {
  const [adressColumnNames, setAdressColumnNames] = useState([])
  const { percent, setProgress } = useProgressStatus(tableData.length)

  function renameLatLongColumns(event) {
    event.preventDefault()

    const latName = event.target.latitude.value
    const longName = event.target.longitude.value

    tableData.forEach((dataset) => {
      if (latName !== 'latitude') {
        dataset.latitude = dataset[latName]
        delete dataset[latName]
      }

      if (longName !== 'longitude') {
        dataset.longitude = dataset[longName]
        delete dataset[longName]
      }
    })

    setMapData(tableData)
  }

  function onChangeAdressTags(event) {
    const tagName = event.target.name
    let newArray = []
    if (adressColumnNames.find((name) => name === tagName)) {
      newArray = adressColumnNames.filter((name) => name !== tagName)
      setAdressColumnNames(newArray)
    } else setAdressColumnNames([...adressColumnNames, tagName])
  }

  async function geocodeAdress(event) {
    event.preventDefault()

    const limiter = new Bottleneck({
      minTime: 100,
    })

    Promise.all(
      tableData.map((dataset, index) => {
        return limiter.schedule(() => {
          setProgress(index)
          return fetchAdress(dataset, adressColumnNames)
        })
      })
    ).then((dataWithCoordinates) => setMapData(dataWithCoordinates))
  }

  return (
    <>
      {percent > 0 ? (
        <>
          <LoadingDots />
          <CenterParagraph>{percent}%</CenterParagraph>
        </>
      ) : (
        <>
          <Form onSubmit={renameLatLongColumns}>
            <p>
              W채hle dazu die Spalten aus, die Latitude und Longitude
              repr채sentieren:
            </p>
            <Label>
              <span>Breitengrad (Latitude):</span>
              <select name="latitude">
                {columnNames.map((name) => (
                  <option key={name} value={name}>
                    {name}
                  </option>
                ))}
              </select>
            </Label>
            <Label>
              <span>L채ngengrad (Longitude):</span>
              <select name="longitude">
                {columnNames.map((name) => (
                  <option key={name} value={name}>
                    {name}
                  </option>
                ))}
              </select>
            </Label>

            <StyledButton type="submit" styleType="more">
              Karte erstellen
            </StyledButton>
          </Form>
          <CenterParagraph>
            <b>ODER</b>
          </CenterParagraph>
          <Form onSubmit={geocodeAdress}>
            <p>W채hle alle Spaltenamen, die ein Teil der Adresse beinhalten:</p>
            <div>
              {columnNames.map((name) => (
                <Tag
                  key={name}
                  name={name}
                  onClick={onChangeAdressTags}
                  type="button"
                  active={adressColumnNames.find((columnName) =>
                    columnName === name ? true : false
                  )}
                >
                  {name}
                </Tag>
              ))}
            </div>
            <StyledButton type="submit" styleType="more">
              Karte erstellen
            </StyledButton>
          </Form>
        </>
      )}
    </>
  )
}

const Tag = styled.button`
  color: ${(props) => (props.active ? 'white' : 'var(--grey)')};
  background: ${(props) => (props.active ? 'var(--cyan)' : 'var(--lightgrey)')};
  border-radius: 15px;
  font-size: 12px;
  padding: 15px 10px;
  margin: 2px;
  text-align: center;
  cursor: pointer;
  border: none;
  font-weight: 700;

  :hover {
    border: 1px solid var(--cyan);
    padding: 14px 9px;
  }
`

const Form = styled.form`
  margin: 30px 0 30px 0;
  border: 1px solid var(--grey);
  padding: 20px;
  border-radius: 5px;
  display: flex;
  flex-direction: column;
`

const Label = styled.label`
  display: block;
  margin: 0 0 10px 0;
`

const StyledButton = styled(Button)`
  align-self: center;
  margin: 15px;
`

const CenterParagraph = styled.p`
  place-self: center;
`
