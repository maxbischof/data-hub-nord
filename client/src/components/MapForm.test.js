import React from 'react'
import MapForm from './MapForm'
import renderer from 'react-test-renderer'

const tableData = [
  {
    id: '9092146',
    name: '2. Polizeirevier Neumünster',
    city: 'Neumünster',
    zipcode: '24536',
    street: 'Hansaring',
    houseNumber: '122',
    telephone: '+49 4321 945-1211',
    fax: '+49 4321 945-1205',
    email: 'Neumünster.PRev@polizei.landsh.de',
    website: '',
    longitude: '9.966029127405152',
    latitude: '54.076095001735126',
    '.': '.',
  },
  {
    id: '9092250',
    name: '3. Polizeirevier Kiel',
    city: 'Kiel',
    zipcode: '24114',
    street: 'Von-der-Tann-Straße',
    houseNumber: '34',
    telephone: '+49 431 160-1310',
    fax: '+49 431 160-1319',
    email: 'Kiel.Prev03@polizei.landsh.de',
    website: '',
    longitude: '10.123243886338098',
    latitude: '54.31637046286201',
    '.': '.',
  },
  {
    id: '9092407',
    name: '3. Polizeirevier Lübeck',
    city: 'Lübeck, Hansestadt',
    zipcode: '23566',
    street: 'Meesenring',
    houseNumber: '9',
    telephone: '+49 451 131-6345',
    fax: '+49 451 131-6319',
    email: 'luebeck-prev03@polizei.landsh.de',
    website: '',
    longitude: '10.71990326564793',
    latitude: '53.86511320627785',
    '.': '.',
  },
  {
    id: '',
  },
]

const columnNames = tableData && Object.keys(tableData[0])
const mapData = null
function setMapData(data) {
  mapData = data
}

it('renders map form component', () => {
  const renderedTable = renderer
    .create(
      <MapForm
        tableData={tableData}
        columnNames={columnNames}
        setMapData={setMapData}
      />
    )
    .toJSON()
  expect(renderedTable).toMatchSnapshot()
})
