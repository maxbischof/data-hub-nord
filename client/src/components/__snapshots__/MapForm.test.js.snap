// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`renders map form component 1`] = `
Array [
  <form
    className="sc-dlnjPT faXGof"
    onSubmit={[Function]}
  >
    <p>
      W채hle dazu die Spalten aus, die Latitude und Longitude repr채sentieren:
    </p>
    <label
      className="sc-hKFyIo eWhztp"
    >
      <span>
        Breitengrad (Latitude):
      </span>
      <select
        name="latitude"
      >
        <option
          value="id"
        >
          id
        </option>
        <option
          value="name"
        >
          name
        </option>
        <option
          value="city"
        >
          city
        </option>
        <option
          value="zipcode"
        >
          zipcode
        </option>
        <option
          value="street"
        >
          street
        </option>
        <option
          value="houseNumber"
        >
          houseNumber
        </option>
        <option
          value="telephone"
        >
          telephone
        </option>
        <option
          value="fax"
        >
          fax
        </option>
        <option
          value="email"
        >
          email
        </option>
        <option
          value="website"
        >
          website
        </option>
        <option
          value="longitude"
        >
          longitude
        </option>
        <option
          value="latitude"
        >
          latitude
        </option>
        <option
          value="."
        >
          .
        </option>
      </select>
    </label>
    <label
      className="sc-hKFyIo eWhztp"
    >
      <span>
        L채ngengrad (Longitude):
      </span>
      <select
        name="longitude"
      >
        <option
          value="id"
        >
          id
        </option>
        <option
          value="name"
        >
          name
        </option>
        <option
          value="city"
        >
          city
        </option>
        <option
          value="zipcode"
        >
          zipcode
        </option>
        <option
          value="street"
        >
          street
        </option>
        <option
          value="houseNumber"
        >
          houseNumber
        </option>
        <option
          value="telephone"
        >
          telephone
        </option>
        <option
          value="fax"
        >
          fax
        </option>
        <option
          value="email"
        >
          email
        </option>
        <option
          value="website"
        >
          website
        </option>
        <option
          value="longitude"
        >
          longitude
        </option>
        <option
          value="latitude"
        >
          latitude
        </option>
        <option
          value="."
        >
          .
        </option>
      </select>
    </label>
    <button
      className="sc-bdnylx gzvFCc sc-eCApGN dvpocg"
      onClick={[Function]}
    >
      Karte erstellen
    </button>
  </form>,
  <p
    className="sc-jSFkmK dTIjFX"
  >
    <b>
      ODER
    </b>
  </p>,
  <form
    className="sc-dlnjPT faXGof"
    onSubmit={[Function]}
  >
    <p>
      W채hle alle Spaltenamen, die ein Teil der Adresse beinhalten:
    </p>
    <div>
      <button
        className="sc-gtssRu bbTUDT"
        name="id"
        onClick={[Function]}
        type="button"
      >
        id
      </button>
      <button
        className="sc-gtssRu bbTUDT"
        name="name"
        onClick={[Function]}
        type="button"
      >
        name
      </button>
      <button
        className="sc-gtssRu bbTUDT"
        name="city"
        onClick={[Function]}
        type="button"
      >
        city
      </button>
      <button
        className="sc-gtssRu bbTUDT"
        name="zipcode"
        onClick={[Function]}
        type="button"
      >
        zipcode
      </button>
      <button
        className="sc-gtssRu bbTUDT"
        name="street"
        onClick={[Function]}
        type="button"
      >
        street
      </button>
      <button
        className="sc-gtssRu bbTUDT"
        name="houseNumber"
        onClick={[Function]}
        type="button"
      >
        houseNumber
      </button>
      <button
        className="sc-gtssRu bbTUDT"
        name="telephone"
        onClick={[Function]}
        type="button"
      >
        telephone
      </button>
      <button
        className="sc-gtssRu bbTUDT"
        name="fax"
        onClick={[Function]}
        type="button"
      >
        fax
      </button>
      <button
        className="sc-gtssRu bbTUDT"
        name="email"
        onClick={[Function]}
        type="button"
      >
        email
      </button>
      <button
        className="sc-gtssRu bbTUDT"
        name="website"
        onClick={[Function]}
        type="button"
      >
        website
      </button>
      <button
        className="sc-gtssRu bbTUDT"
        name="longitude"
        onClick={[Function]}
        type="button"
      >
        longitude
      </button>
      <button
        className="sc-gtssRu bbTUDT"
        name="latitude"
        onClick={[Function]}
        type="button"
      >
        latitude
      </button>
      <button
        className="sc-gtssRu bbTUDT"
        name="."
        onClick={[Function]}
        type="button"
      >
        .
      </button>
    </div>
    <button
      className="sc-bdnylx gzvFCc sc-eCApGN dvpocg"
      onClick={[Function]}
    >
      Karte erstellen
    </button>
  </form>,
]
`;
